pkg/scheduler/queue/queue_test.go

```diff
@@ -845,3 +865,142 @@ func TestRequestQueue_tryDispatchRequestToQuerier_ShouldReEnqueueAfterFailedSend
+			log.NewNopLogger(),
+			1,
+			0,
+			promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"user"}),
+			promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"user"}),
+			promauto.With(nil).NewHistogram(prometheus.HistogramOpts{}),
+			inflight,
+			promauto.With(nil).NewSummaryVec(prometheus.SummaryOpts{}, []string{"query_component"}),
+			queueStopTimeout,
+		)
+		require.NoError(t, err)
+		require.NoError(t, services.StartAndAwaitRunning(ctx, queue))
+
+		// And then record a request as inflight in the scheduler, even though we're not putting anything in the queue
+		inflight.Add(1)
+
+		// Then stop the Queue
+		start := time.Now()
+		assert.NoError(t, services.StopAndAwaitTerminated(ctx, queue))
+
+		// And ensure that it waits until the timeout to exit (using synctest.Wait so we're not actually waiting)
+		synctest.Wait()
+		require.LessOrEqual(t, queueStopTimeout, time.Since(start))
+		require.LessOrEqual(t, time.Since(start), queueStopTimeout+time.Second)
+	})
+}
+
+// This test ensures that even if the queue has no pending requests, we still wait until any inflight requests
+// have been returned before existing
+func TestRequestQueue_ShutdownWithInflightRequests_ShouldDrainRequests(t *testing.T) {
+	ctx := context.Background()
+	tenantID := "testTenant"
+	querierID := "querier1"
+
+	// We care about the inflight requests
+	inflight := atomic.NewInt64(0)
+
+	// So we create a queue
+	queue, err := NewRequestQueue(
+		log.NewNopLogger(),
+		1,
+		0,
+		promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"user"}),
+		promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"user"}),
+		promauto.With(nil).NewHistogram(prometheus.HistogramOpts{}),
+		inflight,
+		promauto.With(nil).NewSummaryVec(prometheus.SummaryOpts{}, []string{"query_component"}),
+		queueStopTimeout,
+	)
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(ctx, queue))
+
+	// With a worker
+	conn := NewUnregisteredQuerierWorkerConn(context.Background(), querierID)
+	require.NoError(t, queue.AwaitRegisterQuerierWorkerConn(conn))
+
+	// And push a request to the queue
+	req := makeSchedulerRequest(tenantID, []string{})
+	require.NotNil(t, req)
+	err = queue.SubmitRequestToEnqueue(tenantID, req, 1, func() {})
+	require.NoError(t, err)
+
+	// And make sure it got to the queue
+	require.Equal(t, queue.queueBroker.tree.ItemCount(), 1)
+	// Then we record it as inflight
+	require.Equal(t, inflight.Add(1), int64(1))
+
+	// Stop the Queue
+	start := time.Now()
+	queue.StopAsync()
+
+	// Consume the existing request from the queue
+	dequeueReq := NewQuerierWorkerDequeueRequest(conn, FirstTenant())
+	r, _, err := queue.AwaitRequestForQuerier(dequeueReq)
+	require.NoError(t, err)
+	require.NotNil(t, r)
+
+	// Ensure the request has been removed from the queue and remove it from inflight tracking
+	require.Equal(t, queue.queueBroker.tree.ItemCount(), 0)
+	require.Equal(t, inflight.Sub(1), int64(0))
```

Comment:
What happens if we don't directly manipulate the counter per Cursors comment: https://github.com/grafana/mimir/pull/13603/changes#r2570496607 ?